File: php/tests/CommonsTest.php
<?php

namespace Shapes\Tests;

use PHPUnit\Framework\Attributes\TestDox;
use PHPUnit\Framework\TestCase;
use Shapes\Base\Commons\Dimensional;
use Shapes\Tests\Fixtures as Helper;

/**
 * @internal
 *
 * @coversNothing
 */
#[TestDox("Commons Test")]
final class CommonsTest extends TestCase {
    #[TestDox("Test Circle Implements Dimensional")]
    public function testCircleImplementsDimensional(): void {
        $circle = Helper\setup_circle();
        $this->assertDimensional($circle, 754.767, 97.389);
    }

    #[TestDox("Test Rectangle Implements Dimensional")]
    public function testRectangleImplementsDimensional(): void {
        $rectangle = Helper\setup_rectangle();
        $this->assertDimensional($rectangle, 756.437, 112.0);
    }

    #[TestDox("Test Square Implements Dimensional")]
    public function testSquareImplementsDimensional(): void {
        $square = Helper\setup_square();
        $this->assertDimensional($square, 105.062, 41.0);
    }

    private function assertDimensional(
        Dimensional $dimensionalShape,
        float $expectedArea,
        float $expectedPerimeter
    ): void {
        $area_diff = abs($dimensionalShape->area() - $expectedArea);
        $perimeter_diff = abs($dimensionalShape->perimeter() - $expectedPerimeter);

        self::assertTrue($area_diff < 0.001, "AREA NOT EQUAL");
        self::assertTrue($perimeter_diff < 0.001, "PERIMETER NOT EQUAL");
    }
}

File: rust/tests/commons_test.rs
mod fixtures {
    pub mod helper;
}

use fixtures::helper;

mod dimensional_test {
    use shapes::base::commons::Dimensional;

    use super::*;

    #[test]
    fn test_circle_implements_dimensional() {
        let circle = helper::setup_circle();
        assert_dimensional(circle, 754.767, 97.389);
    }

    #[test]
    fn test_rectangle_implements_dimensional() {
        let rectangle = helper::setup_rectangle();
        assert_dimensional(rectangle, 756.437, 112.0);
    }

    #[test]
    fn test_square_implements_dimensional() {
        let square = helper::setup_square();
        assert_dimensional(square, 105.062, 41.0);
    }

    fn assert_dimensional(
        dimensional_shape: impl Dimensional,
        expected_area: f64,
        expected_perimeter: f64,
    ) {
        let area_diff = (dimensional_shape.area() - expected_area).abs();
        let perimeter_diff = (dimensional_shape.perimeter() - expected_perimeter).abs();

        assert!(area_diff < 0.001, "AREA NOT EQUAL");
        assert!(perimeter_diff < 0.001, "PERIMETER NOT EQUAL!");
    }
}
